### PR TITLE
@singledispatch and @singledispatchmethod checks for methods and functions

### PR DESCRIPTION
**Pull request by ramonsaraiva**
## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :sparkles: New feature |

## Description

Draft implementation for `singledispatch` and `singledispatchmethod` checks. ~~I just added one sample functional test file, still need to write the tests for `singledispatchmethod` decorators.~~

~~I don't love the way all those if statements are written~~, but I just want to have a sense if I'm going in the right direction so I can think about cleaning those up. This one function is also checking multiple states:~~
* `singledispatch` with `classmethod`
* `singledispatch` register call with `classmethod`
* `singledispatchmethod` with `staticmethod`
* `singledispatchmethod` register call with `staticmethod`
* `singledispatch` without any other decorator, but still in a class scope

~~I'm pretty sure there's a lot that can be simplified, like moving the lines where I check if the `register` call/attribute actually refers to a `singledispatch` or `singledispatchmethod` decorator to a utility function similar to `is_registered_in_singledispatch_function`. I also realized there is this `decorated_with` helper after I implemented the `decorators_names_set` and `decorators_name_node_map` approach for verifying the cases without `register` calls.~~

EDIT: I think https://github.com/PyCQA/pylint/pull/7575/commits/4d95d21ad41e1a219cc5917181b102e5de5832bd and https://github.com/PyCQA/pylint/pull/7575/commits/cc64ded7c6c5e5fd053e5a58fbbeebce0cff93c8 bring a much cleaner version that https://github.com/PyCQA/pylint/pull/7575/commits/3dc361624cd805e32972372257e3ee48ef1ac84a had - still open to suggestions.

I added a new util function for `singledispatchmethod` (similar to the `singledispatch` one) and moved some reusable logic to a separate one. There was a minor modification to the way we check for `register` calls, cause now with annotated types `register` can also be an attribute and not strictly a call. 

Refs #6917

### CODE REVIEW

File: pylint/checkers/stdlib.py
Comment (DanielNoord) [inline pylint/checkers/stdlib.py:392]: I'd prefer to use a number that is not being used yet. We have `W1510` which at some point we might want to escalate to `E1510`. I believe 1519 and 1520 are free.

File: pylint/checkers/stdlib.py
Comment (DanielNoord) [inline pylint/checkers/stdlib.py:628]:
```suggestion
        for decorator in node.decorators:
```

We check for this in the calling code.

Comment (DanielNoord) [inline pylint/checkers/stdlib.py:628]: We could use `inference` here as users might overwrite `classmethod`. But I think in this case we probably shouldn't. Overwriting `exit` or `print` I can sort of understand but if you're overwriting `classmethod` you probably also don't use `pylint` to make sure you have sane code :sweat_smile:

File: pylint/checkers/stdlib.py
Comment (DanielNoord) [inline pylint/checkers/stdlib.py:641]: Could you add `confidence` as well?

File: pylint/checkers/utils.py
Comment (DanielNoord) [inline pylint/checkers/utils.py:1461]: Can we use `safe_inference` here?

File: pylint/checkers/utils.py
Comment (DanielNoord) [inline pylint/checkers/utils.py:1490]: Perhaps put the refactoring in another PR. That would help reviewing this and keep the size of the PR smaller.

File: tests/functional/s/singledispatch_method.py
Comment (DanielNoord) [inline tests/functional/s/singledispatch_method.py:3]:
```suggestion

# pylint: disable=missing-class-docstring,missing-function-docstring,too-few-public-methods
```

File: tests/functional/s/singledispatchmethod_function.py
Comment (DanielNoord) [inline tests/functional/s/singledispatchmethod_function.py:3]:
```suggestion

# pylint: disable=missing-class-docstring,missing-function-docstring,disable=too-few-public-methods
```

**Review by DanielNoord — CHANGES_REQUESTED** [review]
This already looks very good. Could you also add a `good.py` and `bad.py` example for both messages in the `docs/data` directory?

File: pylint/checkers/utils.py
Comment (ramonsaraiva) [reply pylint/checkers/utils.py:1490]: Just to make sure I understand, by refactoring you mean the common code i moved to `find_infered_fn_from_register`? or not touching `pylint/checkers/utils.py` at all and add the logic for resolving `singledispatch` and `singledispatchmethod` in the stdlib checker module?

Comment (DanielNoord) [reply pylint/checkers/utils.py:1490]: Yeah, the first thing. Let's focus this PR on everything that is strictly necessary for the new messages. Anything else can be moved to separate PRs as it can be easily approved and merged.

File: pylint/checkers/stdlib.py
Comment (ramonsaraiva) [reply pylint/checkers/stdlib.py:392]: Yeah, this was me misunderstanding that numbers are shared between classifiers (W, E, ..)

File: pylint/checkers/utils.py
Comment (ramonsaraiva) [inline pylint/checkers/utils.py:1466]: Still needed to adjust this part cause we should consider, for..
```python
@singledispatch
def example(arg):
    pass
```

..that both below are classified as possible registers:
```python
@example.register(str)
def _(arg):
    pass

@example.register
def _(arg: str):
    pass
```

**Comment by ramonsaraiva** [comment]
For some reason `decorated_with` in py37 is not infering decorator nodes properly so `utils.is_registered_in_singledispatch_function` and `utils.is_registered_in_singledispatchmethod_function` are both returning `False`, which makes this test case to fail:
```
E       Expected in testdata:
E         33: singledispatchmethod-function
E         39: singledispatchmethod-function
```
Only happens on py37, any idea?

**Comment by DanielNoord** [comment]
> For some reason `decorated_with` in py37 is not infering decorator nodes properly so `utils.is_registered_in_singledispatch_function` and `utils.is_registered_in_singledispatchmethod_function` are both returning `False`, which makes this test case to fail:
> 
> ```
> E       Expected in testdata:
> E         33: singledispatchmethod-function
> E         39: singledispatchmethod-function
> ```
> 
> Only happens on py37, any idea?

`singledispatchmethod` was only added in `3.8`. I think that could explain it?

**Comment by ramonsaraiva** [comment]
That does indeed explain.. Sorry for that 🤦🏻‍♂️

File: pylint/checkers/stdlib.py
Comment (DanielNoord) [inline pylint/checkers/stdlib.py:625]: Can we making the mapping a tuple so we only have two temp vars?

I was also thinking of using `if "singledispatch" in decorators_map or "classmethod" in decorators` but I'm not sure that's more readable.

File: tests/functional/s/singledispatch_method.py
Comment (DanielNoord) [inline tests/functional/s/singledispatch_method.py:3]: Which module is reporting this?

Comment (ramonsaraiva) [reply tests/functional/s/singledispatch_method.py:3]: Oh, just realized this only happens for py37 cause `tests/functional/s/singledispatch_method.py` still imports `singledispatchmethod` for testing the right combination of `singledispatchmethod` and `classmethod`. I'll create a py37 and a py38+ versions of that test so I don't do that for py37.

File: pylint/checkers/stdlib.py
Comment (DanielNoord) [inline pylint/checkers/stdlib.py:642]: Perhaps a bit pre-mature but let's skip the temporary assignment here. The code is short and clear enough that we can use `[0]` and `[1]`.

File: pylint/checkers/stdlib.py
Comment (DanielNoord) [inline pylint/checkers/stdlib.py:648]: Same as above.

File: pylint/checkers/utils.py
Comment (DanielNoord) [reply pylint/checkers/utils.py:1466]: Do you need any help with that?

File: pylint/checkers/utils.py
Comment (DanielNoord) [inline pylint/checkers/utils.py:1511]: This should be unreachable based on the typing, right?

**Review by DanielNoord — COMMENTED** [review]
Looking good :smile:

File: pylint/checkers/utils.py
Comment (ramonsaraiva) [reply pylint/checkers/utils.py:1511]: It should, I think I copied this line from the other function that does something similar. I'll remove.

Comment (DanielNoord) [reply pylint/checkers/utils.py:1511]: I mean, I would also be okay with widening the typing and having this accept all types of `nodes`. Although I wouldn't know why you would ever call this with anything other than a `FunctionDef`. Up to you~

**Comment by ramonsaraiva** [comment]
Do you know why that documentation check is failing?
```
sphinx-build -b linkcheck -q -d _build/doctrees -T -W --keep-going   -n . _build/linkcheck
/home/runner/work/pylint/pylint/doc/user_guide/checkers/features.rst:933: WARNING: broken link: https://docs.python.org/3/library/f (404 Client Error: Not Found for url: https://docs.python.org/3/library/f)
```
Here line 933 doesn't have a URL and both functools URLs seem to be complete.

**Comment by DanielNoord** [comment]
@ramonsaraiva You can add the name here: https://github.com/PyCQA/pylint/blob/main/.pyenchant_pylint_custom_dict.txt

**Comment by ramonsaraiva** [comment]
Cool. Besides the documentation check - which I validated that locally the links are not broken - I think we should be good, but please let me know if you think something else needs to be changed.

**Comment by DanielNoord** [comment]
There does seem to be an issue with the documentation. See:

https://pylint--7575.org.readthedocs.build/en/7575/user_guide/checkers/features.html

There is a broken link there. Could you perhaps check if you understand what is causing this?

**Comment by ramonsaraiva** [comment]
@DanielNoord what I found out this morning is that changing `https://docs.python.org/3/library/functools.html#functools.singledispatchmethod` to `https://docs.python.org/3/library/functools.html#functools.singledispatchmet` (<= 90 columns in the line), the doc build doesn't break the URL. Do you have any suggestions on how I can approach this? Maybe we can create an alias in rst for `https://docs.python.org` or something to have those URLs smaller for the auto doc generator?

To avoid increasing scope, I could also remove the reference to the docs and have a followup fixing this case. I also thought about possibly mentioning `See singledispatchmethod in https://docs.python.org/3/library/functools.html`, but not sure if that's as helpful as a direct link to the `singledispatchmethod` documentation.

**Comment by DanielNoord** [comment]
> @DanielNoord what I found out this morning is that changing `https://docs.python.org/3/library/functools.html#functools.singledispatchmethod` to `https://docs.python.org/3/library/functools.html#functools.singledispatchmet` (<= 90 columns in the line), the doc build doesn't break the URL. Do you have any suggestions on how I can approach this? Maybe we can create an alias in rst for `https://docs.python.org` or something to have those URLs smaller for the auto doc generator?
> 
> 
> 
> To avoid increasing scope, I could also remove the reference to the docs and have a followup fixing this case. I also thought about possibly mentioning `See singledispatchmethod in https://docs.python.org/3/library/functools.html`, but not sure if that's as helpful as a direct link to the `singledispatchmethod` documentation.

Either of the two last solutions would be okay for me. I don't really have the bandwidth to follow this up and I don't think the feature should be blocked by something so trivial.

**Review by DanielNoord — DISMISSED** [review]
Thanks @ramonsaraiva!

Reviewing this was a breeze! LGTM but I'll let another maintainer have a look before merging as this is a new checker.

**Comment by ramonsaraiva** [comment]
Thanks for the great guidance @DanielNoord - it was pretty awesome to get at least to the surface of how pylint works. I hope I can help contributing with more things in the near future.

**Review by Pierre-Sassoulas — CHANGES_REQUESTED** [review]
Looks great already, thank you ! Could you add a good/bad example for each messages in the documentation please ? You can take example on https://github.com/PyCQA/pylint/pull/7576

**Comment by ramonsaraiva** [comment]
@Pierre-Sassoulas I tried adding a good and bad example earlier today, but for some reason the `bad.py` was not appearing in the checker page, only the `good.py`. It was setup exactly as the example you sent (https://github.com/PyCQA/pylint/pull/7576). Maybe it was something related to my doc build, but I couldn't get the `bad.py` examples to appear in the documentation page.

**Comment by Pierre-Sassoulas** [comment]
Could you push the work in progress of good/bad please, so we can investigate, please ?

**Comment by ramonsaraiva** [comment]
Will do that early tomorrow, thanks.

File: pylint/checkers/utils.py
Comment (mbyrnepr2) [inline pylint/checkers/utils.py:1483]:
```suggestion
def find_inferred_fn_from_register(node: nodes.NodeNG) -> nodes.FunctionDef | None:
```

File: pylint/checkers/utils.py
Comment (mbyrnepr2) [inline pylint/checkers/utils.py:1513]:
```suggestion
        func_def = find_inferred_fn_from_register(decorator)
```

File: pylint/checkers/utils.py
Comment (mbyrnepr2) [inline pylint/checkers/utils.py:1485]: Docstring instead of comment?

**Review by mbyrnepr2 — COMMENTED** [review]
Overall looks great. A couple of small notes.

File: pylint/checkers/utils.py
Comment (ramonsaraiva) [reply pylint/checkers/utils.py:1485]: There's another similar function that we decided to not yet modify to simplify the review of this pull request, I'll be creating a follow-up pull request that will combine some common functionality between the function that finds `singledispatch` and `singledispatchmethod`. That function also has a similar comment for function calls.

Should I add a proper docstring as part of the followup PR and I'll combine common functionality between both functions?

Comment (mbyrnepr2) [reply pylint/checkers/utils.py:1485]: That sounds good to me 👌

**Comment by ramonsaraiva** [comment]
@Pierre-Sassoulas @mbyrnepr2 https://github.com/PyCQA/pylint/pull/7575/commits/751ae3b95c114b439c3b920de325298b444d59bf adds the `good.py` and `bad.py` examples.. this is what I see in my documentation build after adding them:

image

image

I'm not sure why it is not showing the `bad.py` examples on the page,  I can see that other checkers have both good and bad. LMK if you have any idea of what it could be. Thanks!

**Comment by DanielNoord** [comment]
https://pylint--7575.org.readthedocs.build/en/7575/user_guide/messages/error/singledispatch-method.html

Seems to be working on RTD :sweat_smile:

File: doc/data/messages/s/singledispatch-method/bad.py
Comment (Pierre-Sassoulas) [inline doc/data/messages/s/singledispatch-method/bad.py:4]:
```suggestion
class Board:
```

File: doc/data/messages/s/singledispatch-method/good.py
Comment (Pierre-Sassoulas) [inline doc/data/messages/s/singledispatch-method/good.py:4]:
```suggestion
class Board:
```
Nitpick : if the differences between good/bad are only problematic code construct vs good code construct it's easier to understand the example

File: doc/data/messages/s/singledispatchmethod-function/bad.py
Comment (Pierre-Sassoulas) [inline doc/data/messages/s/singledispatchmethod-function/bad.py:4]:
```suggestion
class Board:
```

File: doc/data/messages/s/singledispatchmethod-function/good.py
Comment (Pierre-Sassoulas) [inline doc/data/messages/s/singledispatchmethod-function/good.py:4]:
```suggestion
class Board:
```

**Review by Pierre-Sassoulas — DISMISSED** [review]
Great first contribution @ramonsaraiva ! This should be available in 2.16.0.

**Review by mbyrnepr2 — APPROVED** [review]

**Comment by ramonsaraiva** [comment]
Thank you all for the support. Really enjoyed contributing here.

**Review by Pierre-Sassoulas — APPROVED** [review]

**Pull request merged**
